File: boj_codes/C++17/15998.cpp
// https://www.acmicpc.net/problem/15998
// First Written : 20210129
// Last Modified : 20210129

#include <cstdio>
#include <cstdlib>
#include <vector>

long long get_gcd(long long a, long long b){
    long long temp;
    while(b != 0){
        temp = a % b;
        a = b;
        b = temp;
    }

    return a;
}

int main(){
    int N;
    long long deposit, balance;
    long long cur_balance = 0;
    const long long max_val = 9000000000000000000;
    long long lower_bound = 0;

    std::vector<long long> candidates;
    scanf("%d", &N);

    while(N--){
        scanf("%lld %lld", &deposit, &balance);

        if(deposit > 0){
            if(cur_balance + deposit != balance){
                printf("-1");
                exit(0);
            }

        }

        else{
            if(cur_balance + deposit != balance){
                candidates.push_back(balance - cur_balance - deposit);

                if(lower_bound < balance)
                    lower_bound = balance;
            }
        }

        cur_balance = balance;
    }

    if(candidates.size() == 0)
        printf("1");
    else{
        long long gcd = candidates[0];

        if(gcd <= lower_bound){
            printf("-1");
            exit(0);
        }

        for(int i = 1; i < candidates.size(); i++){
            gcd = get_gcd(gcd, candidates[i]);

            if(gcd <= lower_bound){
                printf("-1");
                exit(0);
            }

        }

        if(gcd > max_val){
            printf("-1");
        }
        else{
            printf("%lld", gcd);
        }

    }

    return 0;
}

File: boj_codes/Python3/15998.py
# https://www.acmicpc.net/problem/15998
# 20201125

import sys
import math

if __name__ == "__main__":
    N = int(sys.stdin.readline().rstrip())
    MAX_VAL = 9 * 1e18

    prev_balance = 0

    candidates = []
    lower_bound = 0

    for _ in range(N):
        deposit, balance = map(int, sys.stdin.readline().split())

        # If withdrawn and charged
        if deposit < 0 and prev_balance + deposit != balance :

            # Add every M * k
            candidates.append(balance - deposit - prev_balance)

            if lower_bound < balance:
                lower_bound = balance

        # If deposit or not charged
        else:
            if prev_balance + deposit != balance:
                print(-1)
                quit()

        prev_balance = balance

    # If there is no charge
    if len(candidates) == 0:
        print(1)

    else:
        gcd = candidates[0]

        if gcd <= lower_bound:
            print(-1)
            quit()

        for i in range(1, len(candidates)):
            gcd = math.gcd(gcd, candidates[i])

            if gcd <= lower_bound:
                print(-1)
                quit()

        if gcd > MAX_VAL:
            print(-1)

        else:
            print(gcd)

